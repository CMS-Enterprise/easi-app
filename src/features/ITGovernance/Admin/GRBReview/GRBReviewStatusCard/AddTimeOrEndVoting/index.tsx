import React, { useState } from 'react';
import { Controller } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';
import { ErrorMessage } from '@hookform/error-message';
import { Button, ButtonGroup, FormGroup, Label } from '@trussworks/react-uswds';
import {
  GetSystemIntakeGRBReviewDocument,
  SystemIntakeGRBReviewFragment,
  useUpdateSystemIntakeGRBReviewFormInputTimeframeAsyncMutation
} from 'gql/generated/graphql';

import Alert from 'components/Alert';
import DatePickerFormatted from 'components/DatePickerFormatted';
import { useEasiForm } from 'components/EasiForm';
import FieldErrorMsg from 'components/FieldErrorMsg';
import Modal from 'components/Modal';
import RequiredFieldsText from 'components/RequiredFieldsText';
import useMessage from 'hooks/useMessage';
import { formatDateUtc, formatDaysHoursMinutes } from 'utils/date';

type AddTimeOrEndVotingProps = {
  grbReviewAsyncEndDate: SystemIntakeGRBReviewFragment['grbReviewAsyncEndDate'];
  grbVotingInformation: SystemIntakeGRBReviewFragment['grbVotingInformation'];
};

const AddTimeOrEndVoting = ({
  grbReviewAsyncEndDate,
  grbVotingInformation
}: AddTimeOrEndVotingProps) => {
  const { t } = useTranslation('grbReview');

  const { systemId } = useParams<{
    systemId: string;
  }>();

  const { showMessage } = useMessage();

  const [err, setError] = useState<boolean>(false);

  const [isOpen, setIsOpen] = useState<'addTime' | 'endVoting' | null>(null);

  const [mutation] =
    useUpdateSystemIntakeGRBReviewFormInputTimeframeAsyncMutation({
      refetchQueries: [GetSystemIntakeGRBReviewDocument]
    });

  const { days, hours, minutes } = formatDaysHoursMinutes(
    grbReviewAsyncEndDate
  );

  const form = useEasiForm<any>({
    defaultValues: {
      grbReviewAsyncEndDate: ''
    },
    values: {
      grbReviewAsyncEndDate: ''
    }
  });
  const {
    control,
    reset,
    watch,
    handleSubmit,
    formState: { errors }
  } = form;

  const resetModal = () => {
    setError(false);
    reset();
    setIsOpen(null);
  };

  const addTimeOrEndEarly = handleSubmit(async input => {
    mutation({
      variables: {
        input: {
          systemIntakeID: systemId,
          grbReviewAsyncEndDate:
            isOpen === 'endVoting'
              ? new Date().toISOString()
              : input.grbReviewAsyncEndDate
        }
      }
    })
      .then(() => {
        if (isOpen === 'endVoting') {
          showMessage(t('statusCard.endVotingModal.success'), {
            type: 'success'
          });
          return;
        }
        showMessage(
          t('statusCard.addTimeModal.success', {
            date: formatDateUtc(input.grbReviewAsyncEndDate, 'MM/dd/yyyy')
          }),
          {
            type: 'success'
          }
        );

        resetModal();
      })
      .catch(() => {
        setError(true);
      });
  });

  return (
    <>
      {/* ADD TIME MODAL */}
      <Modal
        isOpen={isOpen === 'addTime'}
        closeModal={() => resetModal()}
        className="easi-body-normal padding-bottom-1 overflow-visible width-mobile-lg"
      >
        <h3 className="margin-top-0 margin-bottom-0">
          {t('statusCard.addTimeModal.heading')}
        </h3>

        {err && (
          <Alert type="error" slim>
            {t('technicalAssistance:documents.upload.error')}
          </Alert>
        )}

        <p>{t('statusCard.addTimeModal.description')}</p>

        <RequiredFieldsText className="margin-top-0 font-body-sm" />

        <Controller
          control={control}
          name="grbReviewAsyncEndDate"
          render={({ field: { ref, ...field }, fieldState: { error } }) => (
            <>
              <FormGroup className="margin-top-0">
                <Label htmlFor="grbReviewAsyncEndDate" requiredMarker>
                  {t('statusCard.addTimeModal.label')}
                </Label>

                <p
                  className="text-base-dark margin-y-1"
                  style={{
                    whiteSpace: 'pre-wrap'
                  }}
                >
                  {t('statusCard.addTimeModal.hint')}
                </p>

                <ErrorMessage
                  errors={errors}
                  name="grbReviewAsyncEndDate"
                  as={<FieldErrorMsg />}
                />

                <DatePickerFormatted
                  id="grbReviewAsyncEndDate"
                  {...field}
                  dateInPastWarning
                  value={field.value ?? ''}
                  onChange={date => {
                    if (date !== field.value) {
                      field.onChange(date || ''); // Only update when there's a change
                    }
                  }}
                />

                {!!watch('grbReviewAsyncEndDate') && (
                  <p className="text-base">
                    {t('statusCard.addTimeModal.newEnd', {
                      date: formatDateUtc(
                        watch('grbReviewAsyncEndDate'),
                        'MM/dd/yyyy'
                      )
                    })}
                  </p>
                )}
              </FormGroup>
            </>
          )}
        />

        <ButtonGroup className="margin-top-4">
          <Button
            type="button"
            disabled={!watch('grbReviewAsyncEndDate')?.trim()}
            onClick={() => {
              addTimeOrEndEarly();
            }}
          >
            {t('statusCard.addTimeModal.addTime')}
          </Button>

          <Button
            type="button"
            className="margin-left-2 margin-top-2"
            unstyled
            onClick={() => {
              resetModal();
            }}
          >
            {t('statusCard.addTimeModal.goBack')}
          </Button>
        </ButtonGroup>
      </Modal>

      {/* END VOTING MODAL */}
      <Modal
        isOpen={isOpen === 'endVoting'}
        closeModal={() => resetModal()}
        className="easi-body-normal padding-bottom-1 overflow-visible width-mobile-lg"
      >
        <h3 className="margin-top-0 margin-bottom-0">
          {t('statusCard.endVotingModal.heading')}
        </h3>

        {err && (
          <Alert type="error" slim>
            {t('technicalAssistance:documents.upload.error')}
          </Alert>
        )}

        <p>{t('statusCard.endVotingModal.description')}</p>

        <p className="text-bold margin-top-1 margin-bottom-0">
          {t('statusCard.endVotingModal.timeRemaining')}
        </p>

        <p className="easi-body-large margin-top-0 margin-bottom-2 margin-right-2">
          {t('statusCard.countdown', {
            days,
            hours,
            minutes
          })}
        </p>

        <p className="text-bold margin-top-1 margin-bottom-0">
          {t('statusCard.endVotingModal.votingStatus')}
        </p>

        <p className="easi-body-large margin-top-0 margin-bottom-05 margin-right-2">
          {t('decisionCard.voteInfo', {
            noObjection: grbVotingInformation.numberOfNoObjection,
            objection: grbVotingInformation.numberOfObjection,
            notVoted: grbVotingInformation.numberOfNotVoted
          })}
        </p>

        <ButtonGroup className="margin-top-4">
          <Button
            type="button"
            secondary
            onClick={() => {
              addTimeOrEndEarly();
            }}
          >
            {t('statusCard.endVotingModal.endEarly')}
          </Button>

          <Button
            type="button"
            className="margin-left-2 margin-top-2"
            unstyled
            onClick={() => {
              resetModal();
            }}
          >
            {t('statusCard.endVotingModal.goBack')}
          </Button>
        </ButtonGroup>
      </Modal>

      <ButtonGroup>
        <Button
          type="button"
          outline
          onClick={() => {
            setIsOpen('addTime');
          }}
        >
          {' '}
          {t('statusCard.addTime')}
        </Button>

        <Button
          type="button"
          outline
          onClick={() => {
            setIsOpen('endVoting');
          }}
        >
          {t('statusCard.endVoting')}
        </Button>
      </ButtonGroup>
    </>
  );
};

export default AddTimeOrEndVoting;
