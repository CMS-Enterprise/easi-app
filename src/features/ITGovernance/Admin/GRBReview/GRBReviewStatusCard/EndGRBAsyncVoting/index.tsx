import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';
import { Button, ButtonGroup } from '@trussworks/react-uswds';
import {
  SystemIntakeGRBReviewFragment,
  useEndGRBReviewAsyncVotingMutation
} from 'gql/generated/graphql';

import Alert from 'components/Alert';
import Modal from 'components/Modal';
import useMessage from 'hooks/useMessage';
import { formatDaysHoursMinutes } from 'utils/date';

type EndGRBAsyncVotingProps = {
  grbReviewAsyncEndDate: SystemIntakeGRBReviewFragment['grbReviewAsyncEndDate'];
  grbVotingInformation: SystemIntakeGRBReviewFragment['grbVotingInformation'];
};

/**
 * Displays End Voting button that triggers modal to manually end the GRB voting period
 */
const EndGRBAsyncVoting = ({
  grbReviewAsyncEndDate,
  grbVotingInformation
}: EndGRBAsyncVotingProps) => {
  const { t } = useTranslation('grbReview');

  const { systemId } = useParams<{
    systemId: string;
  }>();

  const { showMessage } = useMessage();

  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [err, setError] = useState<boolean>(false);

  const [mutate] = useEndGRBReviewAsyncVotingMutation({
    variables: {
      systemIntakeID: systemId
    }
  });

  const endVoting = () => {
    mutate()
      .then(() =>
        showMessage(t('statusCard.endVotingModal.success'), {
          type: 'success'
        })
      )
      .catch(() => setError(true));
  };

  const { days, hours, minutes } = formatDaysHoursMinutes(
    grbReviewAsyncEndDate
  );

  return (
    <>
      <Modal
        isOpen={isOpen}
        closeModal={() => setIsOpen(false)}
        className="easi-body-normal padding-bottom-1 maxw-mobile-lg height-fit-content"
      >
        <h3 className="margin-top-0 margin-bottom-0">
          {t('statusCard.endVotingModal.heading')}
        </h3>

        {err && (
          <Alert type="error" slim>
            {t('statusCard.endVotingModal.error')}
          </Alert>
        )}

        <p>{t('statusCard.endVotingModal.description')}</p>

        <p className="text-bold margin-top-1 margin-bottom-0">
          {t('statusCard.endVotingModal.timeRemaining')}
        </p>

        <p className="easi-body-medium margin-top-0 margin-bottom-2 margin-right-2">
          {t('statusCard.countdown', {
            days,
            hours,
            minutes
          })}
        </p>

        <p className="text-bold margin-top-1 margin-bottom-0">
          {t('statusCard.endVotingModal.votingStatus')}
        </p>

        <p className="easi-body-medium margin-top-0 margin-bottom-05 margin-right-2">
          {t('decisionCard.voteInfo', {
            noObjection: grbVotingInformation.numberOfNoObjection,
            objection: grbVotingInformation.numberOfObjection,
            notVoted: grbVotingInformation.numberOfNotVoted
          })}
        </p>

        <ButtonGroup className="margin-top-4">
          <Button type="button" secondary onClick={() => endVoting()}>
            {t('statusCard.endVotingModal.endEarly')}
          </Button>

          <Button
            type="button"
            className="margin-left-2"
            unstyled
            onClick={() => setIsOpen(false)}
          >
            {t('statusCard.endVotingModal.goBack')}
          </Button>
        </ButtonGroup>
      </Modal>

      {/* Modal trigger button */}
      <Button
        type="button"
        outline
        onClick={() => {
          setIsOpen(true);
        }}
      >
        {t('statusCard.endVoting')}
      </Button>
    </>
  );
};

export default EndGRBAsyncVoting;
