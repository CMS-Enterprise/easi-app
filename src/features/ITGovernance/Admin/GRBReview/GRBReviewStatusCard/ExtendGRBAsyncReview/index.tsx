import React, { useState } from 'react';
import { Controller } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';
import { ErrorMessage } from '@hookform/error-message';
import { Button, ButtonGroup, FormGroup, Label } from '@trussworks/react-uswds';
import {
  GetSystemIntakeDocument,
  GetSystemIntakeGRBReviewDocument,
  useExtendGRBReviewDeadlineAsyncMutation
} from 'gql/generated/graphql';

import Alert from 'components/Alert';
import DatePickerFormatted from 'components/DatePickerFormatted';
import { useEasiForm } from 'components/EasiForm';
import FieldErrorMsg from 'components/FieldErrorMsg';
import Modal from 'components/Modal';
import RequiredFieldsText from 'components/RequiredFieldsText';
import useMessage from 'hooks/useMessage';
import { formatDateUtc } from 'utils/date';

/**
 * Displays Add Time button that triggers modal to extend the GRB review deadline
 */
const ExtendGRBAsyncReview = () => {
  const { t } = useTranslation('grbReview');

  const { systemId } = useParams<{
    systemId: string;
  }>();

  const { showMessage } = useMessage();

  const [err, setError] = useState<boolean>(false);

  const [isOpen, setIsOpen] = useState<boolean>(false);

  const [mutation] = useExtendGRBReviewDeadlineAsyncMutation({
    refetchQueries: [GetSystemIntakeGRBReviewDocument, GetSystemIntakeDocument]
  });

  const form = useEasiForm<any>({
    defaultValues: {
      grbReviewAsyncEndDate: ''
    },
    values: {
      grbReviewAsyncEndDate: ''
    }
  });
  const {
    control,
    reset,
    watch,
    handleSubmit,
    formState: { errors }
  } = form;

  const resetModal = () => {
    setError(false);
    reset();
    setIsOpen(false);
  };

  const addTimeOrEndEarly = handleSubmit(async input => {
    mutation({
      variables: {
        input: {
          systemIntakeID: systemId,
          grbReviewAsyncEndDate: input.grbReviewAsyncEndDate
        }
      }
    })
      .then(() => {
        showMessage(
          t('statusCard.addTimeModal.success', {
            date: formatDateUtc(input.grbReviewAsyncEndDate, 'MM/dd/yyyy')
          }),
          {
            type: 'success'
          }
        );

        resetModal();
      })
      .catch(() => {
        setError(true);
      });
  });

  return (
    <>
      {/* ADD TIME MODAL */}
      <Modal
        isOpen={isOpen}
        closeModal={() => resetModal()}
        className="easi-body-normal padding-bottom-1 maxw-mobile-lg height-fit-content"
      >
        <h3 className="margin-top-0 margin-bottom-0">
          {t('statusCard.addTimeModal.heading')}
        </h3>

        {err && (
          <Alert type="error" slim>
            {t('statusCard.addTimeModal.error')}
          </Alert>
        )}

        <p>{t('statusCard.addTimeModal.description')}</p>

        <RequiredFieldsText className="margin-top-0 font-body-sm" />

        <Controller
          control={control}
          name="grbReviewAsyncEndDate"
          render={({ field: { ref, ...field }, fieldState: { error } }) => (
            <>
              <FormGroup className="margin-top-0">
                <Label htmlFor="grbReviewAsyncEndDate" requiredMarker>
                  {t('statusCard.addTimeModal.label')}
                </Label>

                <p
                  className="text-base-dark margin-y-1"
                  style={{
                    whiteSpace: 'pre-wrap'
                  }}
                >
                  {t('statusCard.addTimeModal.hint')}
                </p>

                <ErrorMessage
                  errors={errors}
                  name="grbReviewAsyncEndDate"
                  as={<FieldErrorMsg />}
                />

                <DatePickerFormatted
                  id="grbReviewAsyncEndDate"
                  {...field}
                  dateInPastWarning
                  value={field.value ?? ''}
                  onChange={date => {
                    if (date !== field.value) {
                      field.onChange(date || ''); // Only update when there's a change
                    }
                  }}
                />

                {!!watch('grbReviewAsyncEndDate') && (
                  <p className="text-base">
                    {t('statusCard.addTimeModal.newEnd', {
                      date: formatDateUtc(
                        watch('grbReviewAsyncEndDate'),
                        'MM/dd/yyyy'
                      )
                    })}
                  </p>
                )}
              </FormGroup>
            </>
          )}
        />

        <ButtonGroup className="margin-top-4">
          <Button
            type="button"
            data-testid="addTimeModalButton"
            disabled={!watch('grbReviewAsyncEndDate')?.trim()}
            onClick={() => addTimeOrEndEarly()}
          >
            {t('statusCard.addTimeModal.addTime')}
          </Button>

          <Button
            type="button"
            className="margin-left-2"
            unstyled
            onClick={() => resetModal()}
          >
            {t('statusCard.addTimeModal.goBack')}
          </Button>
        </ButtonGroup>
      </Modal>

      {/* Modal trigger button */}
      <Button type="button" outline onClick={() => setIsOpen(true)}>
        {t('statusCard.addTime')}
      </Button>
    </>
  );
};

export default ExtendGRBAsyncReview;
