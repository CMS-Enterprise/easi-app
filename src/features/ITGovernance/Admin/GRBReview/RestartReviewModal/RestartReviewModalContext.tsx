// src/contexts/ModalContext.tsx
import React, { createContext, ReactNode, useContext, useState } from 'react';

interface ModalContextValue {
  isOpen: boolean;
  openModal: () => void;
  closeModal: () => void;
}

const ModalContext = createContext<ModalContextValue | undefined>(undefined);

interface ModalProviderProps {
  children: ReactNode;
}

export const ModalProvider: React.FC<ModalProviderProps> = ({ children }) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const openModal = () => setIsOpen(true);
  const closeModal = () => setIsOpen(false);

  return (
    <ModalContext.Provider value={{ isOpen, openModal, closeModal }}>
      {children}
    </ModalContext.Provider>
  );
};

export function useRestartReviewModal(): ModalContextValue {
  const context = useContext(ModalContext);
  if (!context) {
    throw new Error(
      'useRestartReviewModal must be used within a ModalProvider'
    );
  }
  return context;
}
