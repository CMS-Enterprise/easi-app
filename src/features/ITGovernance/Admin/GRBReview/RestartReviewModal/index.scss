// classname used to target specifically the date picker calendar and have it render above the input field
.date-picker-override {
  .usa-date-picker__calendar {
    top: -262px !important;

    @media screen and (max-width: 640px) {
      top: -261px !important;
    }
  }
}
