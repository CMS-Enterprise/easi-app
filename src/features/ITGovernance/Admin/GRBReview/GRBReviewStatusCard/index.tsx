import React, { useContext, useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import { Button, ButtonGroup, Grid, Icon } from '@trussworks/react-uswds';
import classNames from 'classnames';
import {
  SystemIntakeGRBReviewFragment,
  SystemIntakeGRBReviewType
} from 'gql/generated/graphql';
import ITGovAdminContext from 'wrappers/ITGovAdminContext/ITGovAdminContext';

import IconLink from 'components/IconLink';
import UswdsReactLink from 'components/LinkWrapper';
import Tag from 'components/Tag';
import { GRBReviewStatus } from 'types/grbReview';
import { formatDateUtc, formatDaysHoursMinutes } from 'utils/date';

import { useRestartReviewModal } from '../RestartReviewModal/RestartReviewModalContext';

import EndGRBAsyncVoting from './EndGRBAsyncVoting';
import ExtendGRBAsyncReview from './ExtendGRBAsyncReview';

type GRBReviewStatusTagProps = {
  systemIntakeId: string;
  isITGovAdmin: boolean;
  grbReviewType: SystemIntakeGRBReviewType;
  grbReviewStatus: GRBReviewStatus;
};

const GRBReviewStatusTag = ({
  systemIntakeId,
  isITGovAdmin,
  grbReviewType,
  grbReviewStatus
}: GRBReviewStatusTagProps) => {
  const { t } = useTranslation('grbReview');
  const { openModal } = useRestartReviewModal();

  const showRestartButton =
    isITGovAdmin &&
    grbReviewType === SystemIntakeGRBReviewType.ASYNC &&
    grbReviewStatus === 'COMPLETED';

  const showSetupButton = isITGovAdmin && grbReviewStatus === 'NOT_STARTED';

  return (
    <div className={classNames('display-flex flex-align-center')}>
      <h4 className="margin-0 margin-right-1 flex-align-self-center">
        {t('statusCard.reviewStatus')}
      </h4>

      <Tag
        data-testid="async-status"
        className="bg-white text-base-darker font-body-sm flex-align-self-center"
      >
        {t(`statusCard.grbReviewStatus.${grbReviewStatus}`)}
      </Tag>

      {showRestartButton && (
        <Button
          type="button"
          unstyled
          onClick={openModal}
          className="margin-left-3"
        >
          {t('statusCard.restartReview')}
          <Icon.ArrowForward aria-hidden />
        </Button>
      )}

      {showSetupButton && (
        <IconLink
          to={`/it-governance/${systemIntakeId}/grb-review/review-type`}
          className="margin-left-3"
          icon={<Icon.ArrowForward aria-hidden />}
          iconPosition="after"
        >
          {t('adminTask.setUpGRBReview.title')}
        </IconLink>
      )}
    </div>
  );
};

export type GRBReviewStatusCardProps = {
  systemIntakeId: string;
  grbReview: SystemIntakeGRBReviewFragment;
  className?: string;
};

const GRBReviewStatusCard = ({
  systemIntakeId,
  grbReview,
  className
}: GRBReviewStatusCardProps) => {
  const { t } = useTranslation('grbReview');

  const {
    grbReviewType,
    grbReviewStandardStatus,
    grbReviewAsyncStatus,
    grbDate,
    grbReviewStartedAt,
    grbReviewAsyncEndDate,
    grbVotingInformation
  } = grbReview;

  const isITGovAdmin = useContext(ITGovAdminContext);

  const { days, hours, minutes } = formatDaysHoursMinutes(
    grbReviewAsyncEndDate
  );

  /**
   * Returns the correct status data for the review type,
   * or NOT_STARTED if status is null
   */
  const grbReviewStatus: GRBReviewStatus | null | undefined = useMemo(() => {
    if (!grbReviewStartedAt) return 'NOT_STARTED';

    return grbReviewType === SystemIntakeGRBReviewType.STANDARD
      ? grbReviewStandardStatus
      : grbReviewAsyncStatus;
  }, [
    grbReviewAsyncStatus,
    grbReviewStandardStatus,
    grbReviewType,
    grbReviewStartedAt
  ]);

  const backgroundColorClass = useMemo(() => {
    switch (grbReviewStatus) {
      case 'COMPLETED':
        return 'bg-success-lighter';

      case 'NOT_STARTED':
        return 'bg-warning-lighter';

      case 'PAST_DUE':
        return 'bg-error-lighter';

      default:
        return 'bg-primary-lighter';
    }
  }, [grbReviewStatus]);

  const reviewIsInProgress: boolean =
    grbReviewStatus !== 'NOT_STARTED' && grbReviewStatus !== 'COMPLETED';

  if (!grbReviewStatus) {
    return null;
  }

  const borderColorClass =
    grbReviewStatus === 'PAST_DUE'
      ? 'border-error-light'
      : 'border-primary-light';

  const dateTextColorClass =
    grbReviewStatus === 'PAST_DUE' ? 'text-error-dark' : 'text-primary-dark';

  return (
    <div
      className={classNames(
        className,
        'padding-3 radius-md',
        backgroundColorClass
      )}
    >
      {grbReviewStatus !== 'NOT_STARTED' && (
        <h3 className="margin-top-0 margin-bottom-2">
          {t('statusCard.heading', { context: grbReviewType })}
        </h3>
      )}

      {/* Status Section */}
      <GRBReviewStatusTag
        systemIntakeId={systemIntakeId}
        grbReviewType={grbReviewType}
        grbReviewStatus={grbReviewStatus}
        isITGovAdmin={isITGovAdmin}
      />

      {/* Meeting Details */}
      {reviewIsInProgress && (
        <div
          className={classNames(
            'border-top-1px margin-top-2 padding-top-2',
            borderColorClass
          )}
        >
          {grbReviewType === SystemIntakeGRBReviewType.ASYNC ? (
            // ASYNC review type
            <>
              <h4 className="margin-0 margin-right-1 margin-top-2px margin-bottom-05">
                {t('statusCard.timeRemaining')}
              </h4>

              <Grid row className="margin-bottom-05">
                <p className="easi-body-large margin-top-0 margin-bottom-05 margin-right-2">
                  {t('statusCard.countdown', {
                    days,
                    hours,
                    minutes
                  })}
                </p>

                <p
                  className={classNames(
                    'easi-body-normal margin-0 flex-align-self-center',
                    dateTextColorClass
                  )}
                >
                  {t('statusCard.reviewEnds', {
                    date: formatDateUtc(grbReviewAsyncEndDate, 'MM/dd/yyyy'),
                    context: grbReviewStatus
                  })}
                </p>
              </Grid>

              {isITGovAdmin && (
                <ButtonGroup>
                  <ExtendGRBAsyncReview />
                  <EndGRBAsyncVoting
                    grbReviewAsyncEndDate={grbReviewAsyncEndDate}
                    grbVotingInformation={grbVotingInformation}
                  />
                </ButtonGroup>
              )}
            </>
          ) : (
            // STANDARD meeting type
            <>
              <h4 className="margin-0 margin-right-1 margin-top-2px margin-bottom-05">
                {t('statusCard.grbMeeting')}
              </h4>

              <div className="easi-body-large">
                {formatDateUtc(grbDate, 'MM/dd/yyyy')}
              </div>

              {isITGovAdmin && (
                <UswdsReactLink
                  to={`/it-governance/${grbReview.id}/dates`}
                  className="usa-button usa-button--outline margin-top-1"
                >
                  {t('statusCard.changeMeetingDate')}
                </UswdsReactLink>
              )}
            </>
          )}
        </div>
      )}
    </div>
  );
};

export default GRBReviewStatusCard;
