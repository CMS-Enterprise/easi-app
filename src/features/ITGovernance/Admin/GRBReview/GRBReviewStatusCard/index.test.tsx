import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen } from '@testing-library/react';
import {
  GRBVotingInformationStatus,
  SystemIntakeGRBReviewAsyncStatusType,
  SystemIntakeGRBReviewFragment,
  SystemIntakeGRBReviewStandardStatusType,
  SystemIntakeGRBReviewType
} from 'gql/generated/graphql';
import i18next from 'i18next';
import ITGovAdminContext from 'wrappers/ITGovAdminContext/ITGovAdminContext';

import { MessageProvider } from 'hooks/useMessage';

import { ModalProvider } from '../RestartReviewModal/RestartReviewModalContext';

import GRBReviewStatusCard from './index';

describe('GRBReviewStatusCard', () => {
  const mockStandardReview: SystemIntakeGRBReviewFragment = {
    __typename: 'SystemIntake',
    grbReviewType: SystemIntakeGRBReviewType.STANDARD,
    grbReviewAsyncStatus: null,
    grbReviewStandardStatus: SystemIntakeGRBReviewStandardStatusType.SCHEDULED,
    grbDate: '2025-03-28T12:00:00Z',
    grbReviewStartedAt: '2025-03-27T12:00:00Z',
    grbReviewAsyncEndDate: null,
    grbVotingInformation: {
      __typename: 'GRBVotingInformation',
      grbReviewers: [],
      numberOfNoObjection: 0,
      numberOfNotVoted: 0,
      numberOfObjection: 0,
      votingStatus: GRBVotingInformationStatus.NOT_STARTED
    },
    id: '123',
    documents: []
  };

  const mockAsyncReview: SystemIntakeGRBReviewFragment = {
    __typename: 'SystemIntake',
    grbReviewType: SystemIntakeGRBReviewType.ASYNC,
    grbReviewAsyncStatus: SystemIntakeGRBReviewAsyncStatusType.IN_PROGRESS,
    grbDate: null,
    grbReviewStartedAt: '2025-03-27T12:00:00Z',
    grbReviewAsyncEndDate: '2025-03-30T12:00:00Z',
    grbVotingInformation: {
      __typename: 'GRBVotingInformation',
      grbReviewers: [],
      numberOfNoObjection: 0,
      numberOfNotVoted: 0,
      numberOfObjection: 0,
      votingStatus: GRBVotingInformationStatus.NOT_STARTED
    },
    id: '123',
    documents: []
  };

  const renderComponent = (
    grbReview: SystemIntakeGRBReviewFragment,
    isITGovAdmin = true
  ) => {
    return render(
      <MemoryRouter>
        <MockedProvider addTypename={false}>
          <MessageProvider>
            <ModalProvider>
              <ITGovAdminContext.Provider value={isITGovAdmin}>
                <GRBReviewStatusCard grbReview={grbReview} />
              </ITGovAdminContext.Provider>
            </ModalProvider>
          </MessageProvider>
        </MockedProvider>
      </MemoryRouter>
    );
  };

  it('renders the Standard Card when grbReviewType is STANDARD', () => {
    const { asFragment } = renderComponent(mockStandardReview);

    // Check that the standard heading is rendered
    expect(
      screen.getByText(
        i18next.t<string>('grbReview:statusCard.standardHeading')
      )
    ).toBeInTheDocument();

    // Check that the meeting date is displayed
    expect(screen.getByText('03/28/2025')).toBeInTheDocument();

    // Check that the "Change Meeting Date" button is displayed for ITGov admins
    expect(
      screen.getByText(
        i18next.t<string>('grbReview:statusCard.changeMeetingDate')
      )
    ).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the Async Card when grbReviewType is ASYNC', () => {
    renderComponent(mockAsyncReview);

    // Check that the async heading is rendered
    expect(
      screen.getByText(i18next.t<string>('grbReview:statusCard.asyncHeading'))
    ).toBeInTheDocument();

    // Check that the time remaining is displayed
    expect(
      screen.getByText(i18next.t<string>('grbReview:statusCard.timeRemaining'))
    ).toBeInTheDocument();
  });

  it('displays the Add Time or End Voting section for ITGov admins in async review', () => {
    renderComponent(mockAsyncReview);

    // Check that the Add Time or End Voting section is rendered
    expect(
      screen.getByText(i18next.t<string>('grbReview:statusCard.addTime'))
    ).toBeInTheDocument();

    expect(
      screen.getByText(i18next.t<string>('grbReview:statusCard.endVoting'))
    ).toBeInTheDocument();
  });

  it('renders nothing if grbReviewStartedAt is not set', () => {
    const reviewWithoutStartDate = {
      ...mockStandardReview,
      grbReviewStartedAt: null
    };

    const { container } = renderComponent(reviewWithoutStartDate);

    // Ensure the component renders nothing
    expect(container.firstChild).toBeNull();
  });

  it('renders GRB reviewer view - in progress', () => {
    renderComponent(mockStandardReview, false);

    // Check that the Add Time or End Voting button is not displayed
    expect(
      screen.queryByText(i18next.t<string>('grbReview:statusCard.addTime'))
    ).not.toBeInTheDocument();
  });

  it('renders GRB reviewer view - completed', () => {
    const asyncReviewCompletedState = {
      ...mockAsyncReview,
      grbReviewAsyncStatus: SystemIntakeGRBReviewAsyncStatusType.COMPLETED
    };

    renderComponent(asyncReviewCompletedState, false);

    // Hide restart review button
    expect(
      screen.queryByRole('button', { name: /restart review/i })
    ).not.toBeInTheDocument();
  });

  it('renders Completed state in Async reviews', () => {
    const asyncReviewCompletedState = {
      ...mockAsyncReview,
      grbReviewAsyncStatus: SystemIntakeGRBReviewAsyncStatusType.COMPLETED
    };

    renderComponent(asyncReviewCompletedState);

    expect(
      screen.queryByText(
        i18next.t<string>('grbReview:statusCard.restartReview')
      )
    ).toBeInTheDocument();
  });
});
