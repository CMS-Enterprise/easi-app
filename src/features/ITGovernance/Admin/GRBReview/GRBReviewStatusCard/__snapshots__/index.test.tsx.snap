// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GRBReviewStatusCard > Standard meeting card > renders the card 1`] = `
<DocumentFragment>
  <div
    class="padding-3 radius-md bg-primary-lighter"
  >
    <h3
      class="margin-top-0 margin-bottom-2"
    >
      Standard meeting review
    </h3>
    <div
      class="display-flex flex-align-center"
    >
      <h4
        class="margin-0 margin-right-1 flex-align-self-center"
      >
        Review status
      </h4>
      <span
        class="easi-tag bg-white text-base-darker font-body-sm flex-align-self-center"
        data-testid="async-status"
      >
        Scheduled
      </span>
    </div>
    <div
      class="border-top-1px margin-top-2 padding-top-2 border-primary-light"
    >
      <h4
        class="margin-0 margin-right-1 margin-top-2px margin-bottom-05"
      >
        GRB meeting
      </h4>
      <div
        class="easi-body-large"
      >
        03/28/2050
      </div>
      <a
        class="usa-link usa-button usa-button--outline margin-top-1"
        href="/it-governance/123/dates"
      >
        Change meeting date
      </a>
    </div>
  </div>
</DocumentFragment>
`;
